qcodes/tests/parameter/test_delegate_parameter.py

```diff
@@ -512,3 +513,56 @@ def test_underlying_instrument_property_for_delegate_parameter():
 
     d = DelegateParameter('delegate_parameter_without_source', source=None)
     assert d.underlying_instrument is None
+
+
+def test_value_validation():
+    source_param = Parameter("source", set_cmd=None, get_cmd=None)
+    delegate_param = DelegateParameter("delegate", source=source_param)
+
+    delegate_param.vals = vals.Numbers(-10, 10)
+    source_param.vals = None
+    delegate_param.validate(1)
+    with pytest.raises(ValueError):
+        delegate_param.validate(11)
+
+    delegate_param.vals = None
+    source_param.vals = vals.Numbers(-5, 5)
+    delegate_param.validate(1)
+    with pytest.raises(ValueError):
+        delegate_param.validate(6)
+
+    delegate_param.vals = vals.Numbers(-10, 10)
+    source_param.vals = vals.Numbers(-5, 5)
+    delegate_param.validate(1)
+    with pytest.raises(ValueError):
+        delegate_param.validate(6)
+    with pytest.raises(ValueError):
+        delegate_param.validate(11)
+
+
+def test_value_validation_with_offset_and_scale():
+    source_param = Parameter("source", set_cmd=None, get_cmd=None, vals=vals.Numbers(-5, 5))
+    delegate_param = DelegateParameter("delegate", source=source_param, vals=vals.Numbers(-10, 10))
+
+    source_param.offset = 100
+    source_param.scale = None
+    source_param.validate(0)  # raw_value = 100
+    delegate_param.validate(0)  # raw_value = 0
+
+    source_param.offset = None
+    source_param.scale = 100
+    source_param.validate(1)  # raw_value = 100
+    delegate_param.validate(1)  # raw_value = 1
+
+    source_param.offset = None
+    source_param.scale = None
+
+    delegate_param.offset = 100
+    delegate_param.scale = None
+    source_param.validate(0)  # raw_value = 0
+    delegate_param.validate(0)  # raw_value = 100
```

Comment:
sorry, looking at this a bit more i have now a question: shouldn't this validation actually fail? so the delegate parameter has offset of 100, so this means that when the delegate gets set to 0, then the source parameter gets set to 100 (because the delegate parameter passes its raw value to the source parameter, not its value), and this would trigger an error on set, but it does not trigger an error on validate. to confirm my logic or not, could you add `delegate_param.set(0)` to this test and see if it still passes (my guess is that it will fail)?

i think what needs to be done in the implementaion is that validate's call to source.validate should mirror what happens in set_raw where source.set is called or in the delegate cache of DelegateParameter - and that is the call to ``_from_value_to_raw_value(value)``, so the implementation of the validate should actully be:
```python
        super().validate(value)
        if self.source is not None:
            self.source.validate(self._from_value_to_raw_value(value))
```
so if my guess is true, that the ``delegate_param.set(0)`` now fails, then after this change is added to the implementation of validate the ``delegate_param.validate(0)`` would also raise an exception.

Reply:
Yes, I realized this, but I wasn't sure if vals shoud check the raw value or the input value.
From the user point of view, imagine I have a delegate_param to control a DC voltage with 1/1000 voltage divider, i.e. scale=1/1000. If I want to apply 1mV, I can simply use delegate_param(1). Now, to set a safety limit of +/-5mV, should vals be (-5,+5),? or (-5/1000, +5/1000) even though we are using scale=1/1000?

If one looks inside ```DelegateParameter.set```:
```python
def set(self, value: ParamDataType) -> None:
    if self._parameter.source is None:
        raise TypeError(
            "Cannot set the cache of a DelegateParameter "
            "that delegates to None"
        )
    self._parameter.validate(value)
    self._parameter.source.cache.set(
        self._parameter._from_value_to_raw_value(value)
    )
```
it validates the input value, but set the raw_value.
I agree that .validate should somehow mimic .set, but it would change the assignment of vals.

P.S: adding ```delegate_param.set(0)```, the test fails as expected.

Reply:
@astafan8 after thinking about it, I think I understood the problem.
What DelegateParameter.validate should check is the input value using its own validator, and check the raw value with the source's validator.
I am going to implement this in the next commit.

Reply:
> What DelegateParameter.validate should check is the input value using its own validator, and check the raw value with the source's validator.

@Junliang-Wang yes, thank you! now things look correct :)